installinator-common/src/progress.rs

```diff
@@ -2,329 +2,261 @@
-    pub total_elapsed: Duration,
-
-    /// A list of completion events.
-    ///
-    /// Completion events include success and failure events.
-    pub completion_events: Vec<CompletionEvent>,
+#[serde(rename_all = "snake_case")]
+pub enum InstallinatorComponent {
+    /// The host phase 2 component.
+    HostPhase2,
+
+    /// The control plane component.
+    ControlPlane,
+
+    /// A component that means "both the host and the control plane", used for
+    /// writes for now. It is possible that this component will go away in the
+    /// future.
+    Both,
+
+    /// Future variants that might be unknown.
+    #[serde(other, deserialize_with = "deserialize_ignore_any")]
+    Unknown,
+}
 
-    /// A list of progress events.
-    ///
-    /// For now, this is zero or one events since the installinator performs its
-    /// actions serially.
-    pub progress_events: Vec<ProgressEvent>,
+/// Installinator step identifier.
+#[derive(
+    Copy, Clone, Debug, Eq, PartialEq, Deserialize, Serialize, JsonSchema,
+)]
+#[serde(rename_all = "snake_case")]
+pub enum InstallinatorStepId {
+    Download,
+    Format,
+    Write,
 }
 
-/// An individual completion event.
-///
-/// Forms part of [`ProgressReport`].
 #[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize, JsonSchema)]
 #[serde(rename_all = "snake_case", tag = "reason")]
-pub struct CompletionEvent {
-    /// When this event was generated, in duration since the start of the program.
-    ///
-    /// This can be compared against [`ProgressReport::total_elapsed`] to figure
-    /// out how old the event was.
-    pub total_elapsed: Duration,
+pub enum InstallinatorStepMetadata {
+    Write {
+        /// The destination being formatted or written to.
+        ///
+        /// Available with format and destination events.
+        #[schemars(schema_with = "path_schema_opt")]
+        destination: Option<Utf8PathBuf>,
+    },
 
-    /// The kind of event this is.
-    pub kind: CompletionEventKind,
+    /// Future variants that might be unknown.
+    #[serde(other, deserialize_with = "deserialize_ignore_any")]
+    Unknown,
 }
 
-/// An individual kind of completion event.
-///
-/// Forms part of [`CompletionEvent`].
-#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize, JsonSchema)]
 #[serde(rename_all = "snake_case", tag = "reason")]
-pub enum CompletionEventKind {
-    /// The download of an artifact failed.
-    DownloadFailed {
-        /// The download attempt that failed.
-        attempt: usize,
-
-        /// The kind of artifact being downloaded.
-        kind: ArtifactKind,
-
-        /// The peer the artifact was being downloaded from, if any.
+pub enum InstallinatorProgressMetadata {
+    Download {
+        /// The peer being downloaded from.
+        ///
+        /// Available with downlad events.
         peer: SocketAddrV6,
-
```

Comment:
👏 So awesome to collapse these all into `update_engine::ProgressEventKind`

Reply:
Yeah it's so much nicer now.